src/state.py

```diff
@@ -27,15 +32,64 @@ def __init__(self, reason: str):
         self.reason = reason
 
 
+class CharmConfigInvalidError(CharmStateBaseError):
+    """Exception raised when a charm configuration is found to be invalid.
+
+    Attributes:
+        msg: Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the CharmConfigInvalidError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+class ProxyConfig(BaseModel):
+    """Configuration for accessing Jenkins through proxy.
+
+    Attributes:
+        http_proxy: The http proxy URL.
+        https_proxy: The https proxy URL.
+        no_proxy: Comma separated list of hostnames to bypass proxy.
+    """
+
+    http_proxy: typing.Optional[HttpUrl]
+    https_proxy: typing.Optional[HttpUrl]
+    no_proxy: typing.Optional[str]
+
+    @classmethod
+    def from_env(cls) -> typing.Optional["ProxyConfig"]:
+        """Instantiate ProxyConfig from juju charm environment.
+
+        Returns:
+            ProxyConfig if proxy configuration is provided, None otherwise.
+        """
+        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY")
+        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY")
+        no_proxy = os.environ.get("JUJU_CHARM_NO_PROXY")
+        if not http_proxy and not https_proxy:
+            return None
+        # Mypy doesn't understand str is supposed to be converted to HttpUrl by Pydantic.
+        return cls(
+            http_proxy=http_proxy, https_proxy=https_proxy, no_proxy=no_proxy  # type: ignore
+        )
+
+
 @dataclasses.dataclass(frozen=True)
 class State:
     """The tmate-ssh-server operator charm state.
 
     Attributes:
         ip_addr: The host IP address of the given tmate-ssh-server unit.
+        proxy_config: The proxy configuration to apply to services used by tmate.
     """
 
     ip_addr: typing.Union[ipaddress.IPv4Address, ipaddress.IPv6Address, str, None]
+    proxy_config: typing.Optional[ProxyConfig]
```

Comment:
Shouldn't we be a bit more coherent between `ip_addr` which can be None and `proxy_config` which is an Optional ?

Reply:
Fixed, thank you!